lib/instrumentation/modules/@elastic/elasticsearch.js

```diff
@@ -0,0 +1,117 @@
+      super(...args)
+
+      // Mapping an ES client event `result` to its active span.
+      // - Use WeakMap to avoid a leak from possible spans that don't end.
+      // - WeakMap allows us to key off the ES client `request` object itself,
+      //   which means we don't need to rely on `request.id`, which might be
+      //   unreliable because it is user-settable (see `generateRequestId` at
+      //   https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/observability.html)
+      const spanFromEsResult = new WeakMap()
+
+      this.on('request', (err, result) => {
+        const request = result.meta.request
+        agent.logger.debug('intercepted call to @elastic/elasticsearch "request" event %o',
+          { id: request.id, method: request.params.method, path: request.params.path })
+
+        let span = spanFromEsResult.get(result)
+
+        // As of @elastic/elasticsearch@7.10.0-rc.1 this event's `err` will
+        // always be null, but we'll be future-proof.
+        if (err) {
+          agent.captureError(err)
+          if (span !== undefined) {
+            span.end()
+            spanFromEsResult.delete(result)
+          }
+          return
+        }
+
+        // With retries (see `makeRequest` in Transport.js) each attempt will
+        // emit this "request" event using the same `result` object. The
+        // intent is to have one Elasticsearch span plus an HTTP span for each
+        // attempt.
+        if (!span) {
+          span = generateSpan(request.params)
+          if (span) {
+            spanFromEsResult.set(result, span)
+          }
+        }
+        if (!span) {
+          return
+        }
+
+        setElasticsearchDbContext(span, request.params.path,
+          request.params.querystring, request.params.body)
+
+        const { hostname, port } = result.meta.connection.url
+        span.setDestinationContext(
+          getDBDestination(span, hostname, port))
+      })
+
+      this.on('response', (err, result) => {
+        // TODO set "span.outcome" (from err and result.statusCode)
+
+        if (err) {
+          // https://github.com/elastic/elasticsearch-js/blob/master/lib/errors.d.ts
+          // Some of these fields we want to include in the captured error, some
+          // not. Current `captureError` will include all non-object fields.
+          const errOpts = {}
+          if (err.name === 'ResponseError' && err.body && err.body.error) {
+            // Include some data from the Elasticsearch API response body:
+            // https://www.elastic.co/guide/en/elasticsearch/reference/current/common-options.html#common-options-error-options
+            errOpts.custom = {
+              type: err.body.error.type,
+              reason: err.body.error.reason,
+              caused_by: err.body.error.caused_by,
+              status: err.body.status
+            }
+          }
+
+          delete err.data // possibly large or private user data
```

Comment:
might trigger a deopt, cant we just discard it or ignore specific fields from getting captured as part of error instead of deleting them and triggering deopt.

Reply:
@vigneshshanmugam Thanks for bringing it up. I haven't played with tracing deopts at all. Currently the node.js agent's `captureError` doesn't provide a way to guide which error fields should be captured. I can look into adding that capability.

Reply:
https://github.com/elastic/apm-agent-nodejs/pull/1912 is my hoped-for solution for this.

Reply:
That is merged now and this PR moved to using it: https://github.com/elastic/apm-agent-nodejs/pull/1877/commits/20984f22cc073a08d85a551a1b8ddbdf0b312bfd